modules/navier_stokes/src/fvkernels/PCNSFVKT.C

```diff
@@ -0,0 +1,269 @@
+    _pressure_elem(getADMaterialProperty<Real>(NS::pressure)),
+    _pressure_neighbor(getNeighborADMaterialProperty<Real>(NS::pressure)),
+    _grad_pressure_elem(getADMaterialProperty<RealVectorValue>(NS::grad(NS::pressure))),
+    _grad_pressure_neighbor(getNeighborADMaterialProperty<RealVectorValue>(NS::grad(NS::pressure))),
+    _eps_elem(getMaterialProperty<Real>(NS::porosity)),
+    _eps_neighbor(getNeighborMaterialProperty<Real>(NS::porosity)),
+    _eqn(getParam<MooseEnum>("eqn")),
+    _index(getParam<MooseEnum>("momentum_component")),
+    _scalar_elem(_u_elem),
+    _scalar_neighbor(_u_neighbor),
+    _grad_scalar_elem((_eqn == "scalar") ? &_var.adGradSln() : nullptr),
+    _grad_scalar_neighbor((_eqn == "scalar") ? &_var.adGradSlnNeighbor() : nullptr),
+    _limiter(Limiter::build(LimiterType(int(getParam<MooseEnum>("limiter"))))),
+    _knp_for_omega(getParam<bool>("knp_for_omega"))
+{
+  if ((_eqn == "momentum") && !isParamValid("momentum_component"))
+    paramError("eqn",
+               "If 'momentum' is specified for 'eqn', then you must provide a parameter "
+               "value for 'momentum_component'");
+
+#ifndef MOOSE_GLOBAL_AD_INDEXING
+  mooseError(
+      "PCNSFVKT objects require global AD indexing because they  require gradient "
+      "reconstruction. Please cd into the root of your MOOSE directory and run './configure'");
+#endif
+}
+
+std::pair<ADReal, ADReal>
+PCNSFVKT::computeAlphaAndOmega(const ADReal & u_elem_normal,
+                               const ADReal & u_neighbor_normal,
+                               const ADReal & c_elem,
+                               const ADReal & c_neighbor) const
+{
+  // Equations from Greenshields sans multiplication by area (which well be done in
+  // computeResidual/Jacobian
+  const auto psi_elem =
+      std::max({c_elem + u_elem_normal, c_neighbor + u_neighbor_normal, ADReal(0)});
+  const auto psi_neighbor =
+      std::max({c_elem - u_elem_normal, c_neighbor - u_neighbor_normal, ADReal(0)});
+  auto alpha = _knp_for_omega ? psi_elem / (psi_elem + psi_neighbor) : ADReal(0.5);
+  auto omega = _knp_for_omega ? alpha * (1 - alpha) * (psi_elem + psi_neighbor)
+                              : alpha * std::max(psi_elem, psi_neighbor);
+
+  // Do this to avoid new nonzero mallocs
+  const auto dummy_quant = 0 * (c_elem + u_elem_normal + c_neighbor + u_neighbor_normal);
+
+  alpha += dummy_quant;
+  omega += dummy_quant;
+  return std::make_pair(std::move(alpha), std::move(omega));
+}
+
+ADReal
+PCNSFVKT::computeFaceFlux(const ADReal & alpha,
+                          const ADReal & omega,
+                          const ADReal & sup_vel_elem_normal,
+                          const ADReal & sup_vel_neighbor_normal,
+                          const ADReal & adv_quant_elem,
+                          const ADReal & adv_quant_neighbor)
+{
+  return alpha * (sup_vel_elem_normal * adv_quant_elem) +
+         (1 - alpha) * sup_vel_neighbor_normal * adv_quant_neighbor -
```

Comment:
says + in (7) of Greenshields?

Reply:
I'm more used to minus (neighbor - elem) though
Thoughts?

Reply:
Since they're equivalent mathematically I like the way we have it because it looks more like a classic diffusive flux (which it is)